### PR TITLE
feat: add aggregator package

### PR DESCRIPTION
**Pull request by assafgi**

### CODE REVIEW

File: aggregator.go [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+		time.Sleep(5 * time.Minute)
+
+	}
+}
+
+type arrayFlag []string
+
+func (i *arrayFlag) String() string {
+	return "my string representation"
+}
```
Comment (doron-cohen) [inline aggregator.go:110]: Why is this needed?

File: aggregator.go [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+			continue
+		}
+		if new[:i] == prev[prevLen-i:] {
+			break
+		}
+	}
+	return i
+}
+
+func aggregateConsoleOutput(wg *sync.WaitGroup, folder, instanceId string, sess *session.Session) {
```
Comment (doron-cohen) [inline aggregator.go:68]: Better log the start of a new worker for that instance ID

File: aggregator.go [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+
+func main() {
+	log.Println("Running aws console log aggregator service...")
+	var ids arrayFlag
+
+	region := flag.String("region", "eu-central-1", "AWS instance region")
+	folder := flag.String("folder", ".", "Aggregated AWS console log folder")
+	flag.Var(&ids, "id", "List of AWS instances ids")
+	flag.Parse()
+	log.Printf("Region:%s OutputFolder:'%s' InstancesIds:%s\n", *region, *folder, ids)
```
Comment (doron-cohen) [inline aggregator.go]: No need for a newline at the end of each log line: https://golang.org/pkg/log/#Logger.Output

File: aggregator.go [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+	"sync"
+	"time"
+)
+
+func getLatestOutput(instanceId string, sess *session.Session, withLatest bool) string {
+	var consoleInput *ec2.GetConsoleOutputInput
+	if withLatest {
+		consoleInput = &ec2.GetConsoleOutputInput{
+			InstanceId: aws.String(instanceId),
+			Latest:     aws.Bool(true),
```
Comment (doron-cohen) [inline aggregator.go]: Wouldn't `Latest: aws.Bool(withLatest)` make this `if-else` redundant?

File: aggregator.go [hunk 5]
```diff
@@ -0,0 +1,139 @@
+package main
+
+import (
+	"encoding/base64"
```
Comment (doron-cohen) [inline aggregator.go:4]: Is there a reason all indentations are tabs? Not to start a war or something but it is kinda weird...

**Review by doron-cohen — CHANGES_REQUESTED** [review]
A very good job!

I would only add a few simple tests (especially for `findOverlappingEndingIndex`) in a later iteration and add a simple pipeline.

File: aggregator.go [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+	} else {
+		consoleInput = &ec2.GetConsoleOutputInput{
+			InstanceId: aws.String(instanceId),
+		}
+	}
+	ec2Svc := ec2.New(sess)
+
+	result, err := ec2Svc.GetConsoleOutput(consoleInput)
+	if err != nil {
+		if err, ok := err.(awserr.Error); ok {
```
Comment (tigrawap) [inline aggregator.go:24]: not idiomatic
Idiomatic is to handle error in `if; err {`  form
and then continue with "good" flow

In most cases, you just do something and return with error from function
(and that's exactly your case)

File: aggregator.go [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"log"
+	"os"
+	"sync"
+	"time"
+)
+
+func getLatestOutput(instanceId string, sess *session.Session, withLatest bool) string {
```
Comment (tigrawap) [inline aggregator.go]: In general, this function and code overall, would be much much clearer, if you would test instances initially, whether they nitro or not. And then act accordingly
(by instance type, or even by querying this EC2 method. But I'm pretty sure that some API can tell you whether it's nitro or not)

File: aggregator.go [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+			// Non Nitro AWS instance types don't support retrieving the latest console logs
+			if err.Code() == "UnsupportedOperation" && withLatest {
+				return getLatestOutput(instanceId, sess, false)
+			} else {
+				log.Println(err.Error())
+			}
+		} else {
+			log.Println(err.Error())
+		}
+		return ""
```
Comment (tigrawap) [inline aggregator.go]: In general, especially in cases, when function does IOs - it's best to return error. I.e (string, error)

Then you also don't have to print in here, you can print later, in place where you care of errors
This function by itself does not care if it succeeds or not

File: aggregator.go [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+	}
+
+	return string(output)
+}
+
+func findOverlappingEndingIndex(prev, new string) int {
+	var i int
+	prevLen := len(prev)
+	for i = len(new); i > 0; i-- {
+		if i > prevLen {
```
Comment (tigrawap) [inline aggregator.go]: I don't understand, why if the length is bigger then prev length(which is static) it's better to ignore?
And if you want to  ignore, you could as well:
`for i = min(prevLen, newLen); i > 0; i-- {`

File: aggregator.go [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+			continue
+		}
+		if new[:i] == prev[prevLen-i:] {
```
Comment (tigrawap) [inline aggregator.go]: In general. I do not understand what this function is doing and how it's working...
I.e output of the previous call and new call might be totally different

File: aggregator.go (hunk 3 above)
Comment (tigrawap) [inline aggregator.go]: Go code is very boilerplated in general, so when possible and especially when it is easy - best to move to additional function. i.e some parseParams and keep it away from actual logic. But not critical

File: aggregator.go [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+			break
+		}
+	}
+	return i
+}
+
+func aggregateConsoleOutput(wg *sync.WaitGroup, folder, instanceId string, sess *session.Session) {
+	defer wg.Done()
```
Comment (tigrawap) [inline aggregator.go:69]: It maybe does not matter a lot here. As you have nothing to handle
But, in general, it's best to pass context with cancellation callback to this function

And in main - catch signals, invoke cancel on context and test for it inside this function, to exit goroutines in proper manner

**Review by tigrawap — CHANGES_REQUESTED** [review]

File: aggregator.go (hunk 5 above)
Comment (tigrawap) [reply aggregator.go:4]: both ways are acceptible. 
And the good thing about go - there is no place for holy war, `go fmt` resolves all doubts

**Comment by tigrawap** [comment]
Also, you definitely need to have go mod. 
Whatever I had locally did not have Latest flag in sdk

File: aggregator.go (hunk 7 above)
Comment (doron-cohen) [reply aggregator.go]: It would make it cleaner but it will require adding permissions (right now we could run this with only `getConsoleOutput` permission) and maybe even hold a list of Nitro instance types.

File: aggregator.go (hunk 6 above)
Comment (assafgi) [reply aggregator.go:24]: @tigrawap you mean like:
`if result, err := ec2Svc.GetConsoleOutput(consoleInput); err != nil {` ?

File: aggregator.go (hunk 9 above)
Comment (assafgi) [reply aggregator.go]: @tigrawap because I am doing `prev[prevLen-i:]`.
From what I see golang don't have min function for integers: https://www.pixelstech.net/article/1559993656-Why-no-max-min-function-for-integer-in-GoLang
I can implement something like this:

> func min(a, b int) int {
    if a < b {
        return a
    }
    return b
}`

File: aggregator.go (hunk 10 above)
Comment (assafgi) [reply aggregator.go]: The logic is simple.
In each iteration I compare the latest characters of prev to the first characters of new.
Lets say the length of new is`n`. In each iteration I try to compare new `n-i` first characters to prev latest `n-i` characters. If no overlapping found, we will return 0 so we will add all the new output to the log.
Lets say `prev="abcde"` and `new="defg"` I want to add to the log only `fg` so I want to return `2`.

File: aggregator.go (hunk 4 above)
Comment (assafgi) [reply aggregator.go]: @doron-cohen I thought that Non Nitro don't get `Latest` at all. I tested it with `false` value and it looks OK so I will update the code.

File: aggregator.go (hunk 1 above)
Comment (assafgi) [reply aggregator.go:110]: @doron-cohen @tigrawap from what I saw, golang builtin `flag` don't have simple support for list of values : https://stackoverflow.com/questions/28322997/how-to-get-a-list-of-values-into-a-flag-in-golang

File: aggregator.go (hunk 5 above)
Comment (assafgi) [reply aggregator.go:4]: @doron-cohen @tigrawap I am just going to leave this here: https://www.youtube.com/watch?v=SsoOG6ZeyUI

**Review by assafgi — COMMENTED** [review]
@tigrawap do you want me to add `go.mod` to the repo? what about `go.sum`?
seems like it is recommended: https://github.com/golang/go/wiki/Modules#releasing-modules-all-versions

**Comment by tigrawap** [comment]
> @tigrawap do you want me to add `go.mod` to the repo? what about `go.sum`?
> seems like it is recommended: https://github.com/golang/go/wiki/Modules#releasing-modules-all-versions

yes, you should have both to have reliable build

File: aggregator.go (hunk 6 above)
Comment (tigrawap) [reply aggregator.go:24]: Yes, if there is error - you handle it immediately in if, and usually just return from function

Comment (tigrawap) [reply aggregator.go:24]: Good bonus - goland displays such ifs very well

File: aggregator.go (hunk 1 above)
Comment (tigrawap) [reply aggregator.go:110]: and it is by design. Because it is so stupid simple to implement it. And once you do it once for something so simple, you don't think/"afraid" any more of adding new types of flags
Other libraries actually have it built-in. But, it is better to stick with native implementation, because then you can relatively easily pass params to other libraries (like klog/glog log parsers, as they also rely on build-in flags library)

File: aggregator.go (hunk 9 above)
Comment (tigrawap) [reply aggregator.go]: Yeah, in this case, I think they went too far with keeping it stupid simple.
I  think I was adding such min function probably in every go project.
Anyway, whatever you prefer. It's not that big overhead to do many continue's. Pretty minimal overhead. Maybe compiler even optimizes it for you

File: aggregator.go (hunk 10 above)
Comment (tigrawap) [reply aggregator.go]: I have feeling that this logic can break somewhere, but I can't nail it, so probably I'm wrong :)
Anyway, this is elegant.  And unit tests will help to actually understand what you meant

File: aggregator_unit_test.go [hunk 12]
```diff
@@ -0,0 +1,30 @@
+package main
```
Comment (doron-cohen) [inline aggregator_unit_test.go:1]: Usually these files are called `test_{filename}.go` (`test_aggregator.go` in this case)

**Review by doron-cohen — APPROVED** [review]

File: aggregator_unit_test.go (hunk 12 above)
Comment (assafgi) [reply aggregator_unit_test.go:1]: @doron-cohen test files should end with `_test.go`, otherwise `go test` will not find them. See this article: https://medium.com/rungo/unit-testing-made-easy-in-go-25077669318

Comment (doron-cohen) [reply aggregator_unit_test.go:1]: You are right.

**Review by tigrawap — APPROVED** [review]

**Review by tigrawap — APPROVED** [review]
